doc: several documentation improvements

jasnell opened the pull request:
bundled together to make landing them easier.

bsteephenson committed:
docs:events clarify emitter.listener() behavior

Clarifies that emitter.listener() returns a copy, not a reference
Resolves issue #9022

danielrentz committed:
Update util.markdown

Fix for https://github.com/joyent/node/issues/25559 (Typo in example of util.deprecate() documentation)

fresheneesz committed:
doc: clarify Readable._read and Readable.push

Minor clarifications around Readable._read and Readable.push
to make their implementation/usage easier to understand.

https://github.com/joyent/node/issues/14124#issuecomment-115392674

jasnell committed:
doc: two minor stream doc improvements

per: https://github.com/joyent/node/issues/14596

1. document that a runtime error will occur if you attempt
   to unshift after the end event
2. document that calling read after the end event will return
   null and will not trigger a runtime error

jasnell committed:
doc: stream.unshift does not reset reading state

Per https://github.com/joyent/node/issues/14604,

Document that performing an `unshift` during a read
can have unexpected results. Following the `unshift`
with a `push('')` resets the reading state appropriately.
Also indicate that doing an `unshift` during a read
is not optimal and should be avoided.

jasnell commented:
@joyent/node-collaborators

trevnorris commented on doc/api/stream.markdown:
```diff
@@ -167,6 +167,33 @@ again when more data is available.
+var fs = require('fs');
+var rr = fs.createReadStream('foo.txt');
+rr.on('readable', function() {
+  var chunk;
```
Is this variable only for demonstration? Not being used.

jasnell replied:
Ah...right, just mentally zoned on that one ;-). I'll remove the variable really quick

trevnorris commented:
One question. Otherwise LGTM

jasnell committed:
doc: readable event clarification

per https://github.com/joyent/node/issues/14597

Indicate that `'readable'` indicates only that data can
be read from the stream, not that there is actually data
to be consumed. `readable.read([size])` can still return
null. Includes an example that illustrates the point.

chrisdickinson commented on doc/api/stream.markdown:
```diff
@@ -167,6 +167,32 @@ again when more data is available.
 
+Note that the `'readable'` event indicates only that data *can* be
+read from the stream. It does not indicate whether there is actual
+data to be consumed. The callback passed to handle the `'readable'`
```
The first two sentences are pretty confusing. Maybe:

> The `'readable'` event indicates that the stream has new information: either new data is available or the end of the stream has been reached. In the former case, `.read()` will return that data. In the latter case, `.read()` will return `null`. For instance, in the following example, ...

jasnell replied:
:+1: will make that change

jasnell replied:
@chrisdickinson ... done... if it looks good to you I'll get it landed

jasnell committed:
doc: additional refinement to readable event

Per https://github.com/joyent/node/pull/25635#discussion_r33973696

Additional refinement to the clarification on the `readable` event

jasnell commented:
Landed in f91fa52, f7af915, 9faae7a, 423c433, daed421, 340e9f0 and 6036e4f
